### PR TITLE
feat: support ToolQA benchmark

### PR DESCRIPTION
**Pull request by yueqis**
I added support for evaluation of ToolQA benchmark.
ToolQA source repo: https://github.com/night-chen/ToolQA/tree/main.
Gorilla's APIBench arxiv paper: https://arxiv.org/pdf/2306.13304.
All code for evaluation is contained in the directory evaluation/toolqa.
More information on how to use this benchmark is contained in the readme file.

### CODE REVIEW

File: evaluation/toolqa/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# ToolQA Evaluation with OpenDevin
+
+This folder contains evaluation harness we built on top of the original [ToolQA](https://github.com/night-chen/ToolQA) ([paper](https://arxiv.org/pdf/2306.13304)). 
+
+## Setup Environment
+
+Please follow [this document](https://github.com/OpenDevin/OpenDevin/blob/main/Development.md) to setup local develop environment for OpenDevin.
```
Comment (tobitege) [inline evaluation/toolqa/README.md:7]: typo: development

File: evaluation/toolqa/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# ToolQA Evaluation with OpenDevin
+
+This folder contains evaluation harness we built on top of the original [ToolQA](https://github.com/night-chen/ToolQA) ([paper](https://arxiv.org/pdf/2306.13304)).
```
Comment (tobitege) [inline evaluation/toolqa/README.md:3]: an evaluation

File: evaluation/toolqa/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+where `model_config` is mandatory, while all other arguments are optional.
+
+`model_config`, e.g. `llm`, is the config group name for your
+LLM settings, as defined in your `config.toml`.
+
+`agent`, e.g. `CodeActAgent`, is the name of the agent for benchmarks, defaulting
+to `CodeActAgent`.
+
+`eval_limit`, e.g. `10`, limits the evaluation to the first `eval_limit` instances. By
+default, the script evaluates the 1 instance.
```
Comment (tobitege) [inline evaluation/toolqa/README.md:30]: move By to next line

File: evaluation/toolqa/run_infer.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    for process in mp.active_children():
+        print(f'Terminating child process: {process.name}')
+        process.terminate()
+        process.join()
+
+def codeact_user_response(state: State) -> str:
+    msg = (
+        'Please continue working on the task on whatever approach you think is suitable.\n'
+        'When you think you finished the task, respond with `Finish[answer]` where you include your answer in `[]`\n'
+        #'Please run the following command: <execute_bash> exit </execute_bash>.\n'
```
Comment (tobitege) [inline evaluation/toolqa/run_infer.py:40]: remove comment

**Review by tobitege — COMMENTED** [review]
Where is WOLFRAM_APPID used? Is it picked up by litellm?

File: evaluation/toolqa/utils.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+import gdown
+import zipfile
+import os
+import re
+import string
+
+def download_data(dir):
+    data_path = os.path.join(dir, "data/external_corpus")
+    if os.path.exists(data_path): return data_path
+    url = "https://drive.google.com/uc?id=1zRbHzPW2x4dDcfmphBWlan8cxUCRNmqk"
```
Comment (yufansong) [inline evaluation/toolqa/utils.py:15]: Is it safe to expose this url here? Is this a public resource or your personal one?

File: evaluation/toolqa/run_infer.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+import os
+import pathlib
+import subprocess
+import time
+import requests
+from concurrent.futures import ProcessPoolExecutor
+
+import pandas as pd
+import toml
+import whatthepatch
```
Comment (yufansong) [inline evaluation/toolqa/run_infer.py:14]: Seems not used?
```suggestion
```

File: evaluation/toolqa/utils.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+(9) graph_toolkits.check_nodes, which returns the detailed attribute information of a node. You could use this tool by `from tools.graphtools import graph_toolkits`, then `toolkits = graph_toolkits('')`, and then `check_nodes = toolkits.check_nodes`.
+(10) graph_toolkits.check_edges, which returns the detailed attribute information of the edge between two nodes. You could use this tool by `from tools.graphtools import graph_toolkits`, then `toolkits = graph_toolkits('')`, and then `check_edges = toolkits.check_edges`.
+(11) sql_interpreter.execute, which interprets an SQL query and returns the result. You could use this tool by `from tools.sql_interpreter import execute`.
+(12) exec, which interprets some Python code and returns the result.
+
+You may take as many steps as necessary. When you think you finished the task, respond with `Finish[answer]` where you include your answer in `[]`. 
+IMPORTANT: Make sure that in your final answer, you should not print any additional text/instructions other than the actual answer, which should be a word or a simple phrase.
+Question: {question}"""
+
+REACT_INSTRUCTION = """Use tools in the tools directory to solve the task: {question}
```
Comment (yufansong) [inline evaluation/toolqa/utils.py:83]: It seems you define `REACT_INSTRUCTION` twice? And this one will overwrite the previous value.

File: evaluation/toolqa/README.md (hunk 1 above)
Comment (yueqis) [reply evaluation/toolqa/README.md:7]: WOLFRAM_APPID is used in the calculator tool of toolqa.

File: evaluation/toolqa/utils.py (hunk 5 above)
Comment (yueqis) [reply evaluation/toolqa/utils.py:15]: This is public resource. This is the google drive of the data that the ToolQA benchmark requires users to download. (https://github.com/night-chen/ToolQA?tab=readme-ov-file#external-corpus-download)

**Review by yufansong — APPROVED** [review]

**Pull request merged**